src/rust/pants_ng/options/src/pants_invocation.rs

```diff
@@ -0,0 +1,215 @@
+// Copyright 2025 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use std::env;
+use std::iter::Peekable;
+use std::path;
+use std::sync::LazyLock;
+
+use regex::Regex;
+
+static NAME_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^[a-z|A-Z]\w*$").unwrap());
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Args {
+    pub(crate) arg_strings: Vec<String>,
+}
+
+impl Args {
+    // Create an Args instance with the provided args, which must *not* include the
+    // argv[0] process name.
+    pub fn new<I: IntoIterator<Item = String>>(arg_strs: I) -> Self {
+        Self {
+            arg_strings: arg_strs.into_iter().collect(),
+        }
+    }
+
+    pub fn argv() -> Self {
+        let mut args = env::args().collect::<Vec<_>>().into_iter();
+        args.next(); // Consume the process name (argv[0]).
+        // TODO: In Pants's own integration tests we may invoke Pants in a subprocess via
+        //  `python -m pants` or `python path/to/__main__.py` or similar. So
+        //  skipping argv[0] may not be sufficient to get just the set of args to split.
+        //  In practice our tests pass despite these extra args being interpreted as specs
+        //  or goals, but that is skating on thin ice.
+        Self::new(args.collect::<Vec<_>>())
+    }
+}
+
+// Represents a single cli flag used to set an option value, e.g., `--foo`, or `--bar=baz`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Flag {
+    pub key: String, // The flag name, without the `--` prefix, up to the `=` if any.
+    pub value: Option<String>, // The value after the `=`, if any.
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SubCommand {
+    pub name: String,
+    pub flags: Vec<Flag>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Command {
+    pub name: String,
+    pub flags: Vec<Flag>,
+    pub subcommand: Option<SubCommand>,
+}
+
+// The details of a Pants invocation command.
+//
+// An invocation has the form:
+//
+// pants --global-flags \
+//   cmd1 --cmd1-flags (subcmd1 --subcmd1-flags)? \
+//   @ cmd2 --cmd2-flags (subcmd2 --subcmd2-flags)? \
+//   ... \
+//   @ cmdN --cmdN-flags (subcmdN --subcmdN-flags)? \
```

Comment:
Have you considered any other separators like:
- `+`
- `^`
- `';'` (like xargs)

I don't actually like the xargs interface, but it is interesting prior art.